app/src/hid/uhid.h

```diff
@@ -0,0 +1,207 @@
+/* SPDX-License-Identifier: GPL-2.0+ WITH Linux-syscall-note */
+#ifndef __UHID_H_
+#define __UHID_H_
+
+/*
+ * User-space I/O driver support for HID subsystem
+ * Copyright (c) 2012 David Herrmann
+ */
+
+/*
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ */
+
+/*
+ * Public header for user-space communication. We try to keep every structure
+ * aligned but to be safe we also use __attribute__((__packed__)). Therefore,
+ * the communication should be ABI compatible even between architectures.
+ */
+
+#include <stdint.h>
+
+#define BUS_PCI			0x01
+#define BUS_ISAPNP		0x02
+#define BUS_USB			0x03
+#define BUS_HIL			0x04
+#define BUS_BLUETOOTH		0x05
+#define BUS_VIRTUAL		0x06
+
+#define HID_MAX_DESCRIPTOR_SIZE		4096
+
+enum uhid_event_type {
+	__UHID_LEGACY_CREATE,
+	UHID_DESTROY,
+	UHID_START,
+	UHID_STOP,
+	UHID_OPEN,
+	UHID_CLOSE,
+	UHID_OUTPUT,
+	__UHID_LEGACY_OUTPUT_EV,
+	__UHID_LEGACY_INPUT,
+	UHID_GET_REPORT,
+	UHID_GET_REPORT_REPLY,
+	UHID_CREATE2,
+	UHID_INPUT2,
+	UHID_SET_REPORT,
+	UHID_SET_REPORT_REPLY,
+};
+
+struct uhid_create2_req {
+	uint8_t name[128];
+	uint8_t phys[64];
+	uint8_t uniq[64];
+	uint16_t rd_size;
+	uint16_t bus;
+	uint32_t vendor;
+	uint32_t product;
+	uint32_t version;
+	uint32_t country;
+	uint8_t rd_data[HID_MAX_DESCRIPTOR_SIZE];
+} __attribute__((__packed__));
```

Comment:
This is a structure compiled for the computer architecture, it should not be used as is to pass it to the device architecture (in theory, endianness could be different for example).

IMO, the client should only send the descriptor, and the device construct the expected uhid structure on the server side.

Reply:
> IMO, the client should only send the descriptor, and the device construct the expected uhid structure on the server side.

I'd like letting the client to specify device names (like `Scrcpy UHID mouse`/`Scrcpy UHID keyboard`), because the server doesn't know the device's type.

Client also needs to be able to specify `vendor` and `product`, because my gamepad rumble implementation only works with a specific pair of those (by emulating a Bluetooth Xbox One controller to load a special driver).

Because little-endian is so dominant on modern PC and Android, I don't know whether I need to deal with endianness even if those structures are created in Java. The header file is copied from Linux kernel so using C to handle them is the simplest way.

Reply:
> I'd like letting the client to specify device names (like Scrcpy UHID mouse/Scrcpy UHID keyboard), because the server doesn't know the device's type.

Sure. I'd pass all the necessary fields in a control message structure, with serialization/deserialization ((like the other control messages).

Reply:
> I don't know whether I need to deal with endianness even if those structures are created in Java.

I guess so, you must pass a `byte[]` with the expected representation to your `UHidManager.write()` method.
You should probably use a `ByteBuffer` to build it (ref `ByteBuffer.order(…)`, `ByteOrder.nativeOrder()`).

To explain a bit more, I think it is incorrect that the client knows the UHID implementation details (including endianness, some fields that the client does not care about…). Moreover, it is possible to use scrcpy-server standalone (and control it from python for example). Even if we do not absolutely need to consider this usage, it still seems weird to expect to pass a Linux C-struct for UHID messages.

Reply:
I implemented that on [uhid.15](https://github.com/Genymobile/scrcpy/commits/uhid.15/) (cf `UhidManager`):
 - the client sends UHID_CREATE and UHID_INPUT control messages containing HID data
 - the server (device) creates the binary UHID structures to the file descriptor

It works well, but I haven't taken the commits for CAPSLOCK synchronization (HID output) yet.